File: SCAN.py
from functools import cmp_to_key 

ELEVATOR_WEIGHT = 2000 
HUMAN_WEIGHT = 60 
DISTANCE_BETWEEN_FLOORS = 3
ACCELERATION_GRAVITY = 10 

n, cur_level = map(int, input().split( )) 
rq = [] 
for i in range(n): 
    arr = list(map(int, input().split(" ")))
    rq.append(arr) 
left = [] 
right = [] 
for i in range(n): 
    if (rq[i][0] >= cur_level): 
        right.append([i+1, 1])  
    else: 
        left.append([i+1, 1]) 
last = cur_level 
ans = 0 
num = 0 

def cmp(A, B): 
    return rq[A[0]-1][(A[1]+1)%2] - rq[B[0]-1][(B[1]+1)%2] 

while right: 
    right.sort(key = cmp_to_key(cmp)) 
    rq_num = right[0][0]
    if right[0][1] == 1: 
        ans += (num * HUMAN_WEIGHT + ELEVATOR_WEIGHT) * ACCELERATION_GRAVITY * abs(last - rq[rq_num-1][0]) * DISTANCE_BETWEEN_FLOORS
        last = rq[rq_num-1][0] 
        num += 1 
        if rq[rq_num-1][1] >= last: 
            right.append([rq_num, 0])
        else: 
            left.append([rq_num, 0]) 
    elif right[0][1] == 0: 
        ans += (num * HUMAN_WEIGHT + ELEVATOR_WEIGHT) * ACCELERATION_GRAVITY * abs(last - rq[rq_num-1][1]) * DISTANCE_BETWEEN_FLOORS
        last = rq[rq_num-1][1] 
        num -= 1 
    right.pop(0) 

while left: 
    left.sort(key = cmp_to_key(cmp)) 
    rq_num = left[0][0]
    if left[0][1] == 1: 
        ans += (num * HUMAN_WEIGHT + ELEVATOR_WEIGHT) * ACCELERATION_GRAVITY * abs(last - rq[rq_num-1][0]) * DISTANCE_BETWEEN_FLOORS
        last = rq[rq_num-1][0] 
        num += 1 
        if rq[rq_num-1][1] <= last: 
            left.append([rq_num, 0])
        else: 
            right.append([rq_num, 0]) 
    elif left[0][1] == 0: 
        ans += (num * HUMAN_WEIGHT + ELEVATOR_WEIGHT) * ACCELERATION_GRAVITY * abs(last - rq[rq_num-1][1]) * DISTANCE_BETWEEN_FLOORS
        last = rq[rq_num-1][1] 
        num -= 1 
    left.pop(0) 

while right: 
    right.sort(key = cmp_to_key(cmp)) 
    rq_num = right[0][0]
    if right[0][1] == 1: 
        ans += (num * HUMAN_WEIGHT + ELEVATOR_WEIGHT) * ACCELERATION_GRAVITY * abs(last - rq[rq_num-1][0]) * DISTANCE_BETWEEN_FLOORS
        last = rq[rq_num-1][0] 
        num += 1 
    elif right[0][1] == 0: 
        ans += (num * HUMAN_WEIGHT + ELEVATOR_WEIGHT) * ACCELERATION_GRAVITY * abs(last - rq[rq_num-1][1]) * DISTANCE_BETWEEN_FLOORS
        last = rq[rq_num-1][1] 
        num -= 1 
    right.pop(0) 

print(ans)

File: SCAN.cpp
#include <bits/stdc++.h>
using namespace std; 

const int elevatorWeight = 3000;
const int humanWeight = 60;
const int distanceBetweenFloors = 3; 
const int accelerationGravity = 10; 

struct Query{
    int from, to; 
};

vector<Query> rq;

struct cmp1 {
    bool operator() (const pair<int,int>& A, const pair<int,int>& B) {
        int a = A.second == 1 ? rq[A.first-1].from : rq[A.first-1].to;
        int b = B.second == 1 ? rq[B.first-1].from : rq[B.first-1].to; 
        return a > b; 
    }
};

struct cmp2 {
    bool operator() (const pair<int,int>& A, const pair<int,int>& B) {
        int a = A.second == 1 ? rq[A.first-1].from : rq[A.first-1].to; 
        int b = B.second == 1 ? rq[B.first-1].from : rq[B.first-1].to; 
        return a < b; 
    }
};

int main() {
    freopen("input.txt", "r", stdin); 
    int n, curLevel; 
    cin >> n >> curLevel; 
    rq = vector<Query>(n); 
    for (int i = 0; i < n; ++i) {
        cin >> rq[i].from >> rq[i].to; 
    }
    priority_queue<pair<int,int>, vector<pair<int,int>>, cmp2> left; 
    priority_queue<pair<int,int>, vector<pair<int,int>>, cmp1> right; 
    for (int i = 0; i < n; ++i) {
        if (rq[i].from >= curLevel) {
            right.push(make_pair(i+1, 1)); 
        } else {
            left.push(make_pair(i+1, 1)); 
        }
    }
    int last = curLevel; 
    int ans = 0; 
    int num = 0; 
    while (!right.empty()) {
        pair<int,int> cur = right.top(); right.pop();
        // cout << "RIGHT : " << (cur.second == 1 ? rq[cur.first-1].from : rq[cur.first-1].to) << '\n'; 
        // cout << cur.first << ' ' << cur.second << "  "; 
        if (cur.second == 1) {
            ans += (num * humanWeight + elevatorWeight) * accelerationGravity * abs(last - rq[cur.first-1].from) * distanceBetweenFloors; 
            last = rq[cur.first-1].from; 
            ++num; 
            if (rq[cur.first-1].to >= last) {
                right.push(make_pair(cur.first, 0)); 
            } else {
                left.push(make_pair(cur.first, 0)); 
            }
        } else {
            ans += (num * humanWeight + elevatorWeight) * accelerationGravity * abs(last - rq[cur.first-1].to) * distanceBetweenFloors; 
            last = rq[cur.first-1].to; 
            --num; 
        }
    }
    while (!left.empty()) {
        pair<int,int> cur = left.top(); left.pop();
        // cout << "LEFT : " << (cur.second == 1 ? rq[cur.first-1].from : rq[cur.first-1].to) << '\n'; 
        // cout << cur.first << ' ' << cur.second << "  "; 
        if (cur.second == 1) {
            ans += (num * humanWeight + elevatorWeight) * accelerationGravity * abs(last - rq[cur.first-1].from) * distanceBetweenFloors; 
            last = rq[cur.first-1].from; 
            ++num; 
            if (rq[cur.first-1].to <= last) {
                left.push(make_pair(cur.first, 0)); 
            } else {
                right.push(make_pair(cur.first, 0)); 
            }
        } else {
            ans += (num * humanWeight + elevatorWeight) * accelerationGravity * abs(last - rq[cur.first-1].to) * distanceBetweenFloors; 
            last = rq[cur.first-1].to; 
            --num; 
        }
    }
    while (!right.empty()) {
        pair<int,int> cur = right.top(); right.pop();
        // cout << "RIGHT : " << (cur.second == 1 ? rq[cur.first-1].from : rq[cur.first-1].to) << '\n'; 
        // cout << cur.first << ' ' << cur.second << "  "; 
        if (cur.second == 1) {
            ans += (num * humanWeight + elevatorWeight) * accelerationGravity * abs(last - rq[cur.first-1].from) * distanceBetweenFloors; 
            last = rq[cur.first-1].from; 
            ++num; 
        } else {
            ans += (num * humanWeight + elevatorWeight) * accelerationGravity * abs(last - rq[cur.first-1].to) * distanceBetweenFloors; 
            last = rq[cur.first-1].to; 
            --num; 
        }
    }
    // cout << '\n'; 
    cout << ans << '\n'; 

    return 0; 
}
